### PR TITLE
Logging panics in protocolMainLoop 

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#4128

It's not clear to me what's panicking, but this should keep it from crashing the process.

### CODE REVIEW

File: tcpip/transport/tcp/tcp_test.go [hunk 1]
```diff
@@ -269,7 +269,7 @@ func TestTCPResetSentForACKWhenNotUsingSynCookies(t *testing.T) {
 	// are released instantly on Close.
 	tcpTW := tcpip.TCPTimeWaitTimeoutOption(1 * time.Millisecond)
 	if err := c.Stack().SetTransportProtocolOption(tcp.ProtocolNumber, tcpTW); err != nil {
-		t.Fatalf("e.stack.SetTransportProtocolOption(%d, %s) = %s", tcp.ProtocolNumber, tcpTW, err)
+		t.Fatalf("e.stack.SetTransportProtocolOption(%d, %v) = %s", tcp.ProtocolNumber, tcpTW, err)
```
Comment (oxtoacart) [inline tcpip/transport/tcp/tcp_test.go:272]: Test wouldn't compile without these two fixes.

**Comment by max-b** [comment]
I'm not super familiar w/ what's going on here. I can see how `recover`ing from the panic here will address the panic, but is there some other way we'd like to be able to be notified that there's been an error? Is that possible?  Would it make sense to return an error?

**Comment by oxtoacart** [comment]
> I'm not super familiar w/ what's going on here. I can see how `recover`ing from the panic here will address the panic, but is there some other way we'd like to be able to be notified that there's been an error? Is that possible? Would it make sense to return an error?

I was going to do that but then realized that the callers of protocolMainLoop don't actually check the return value. My hope is that we might get some useful logging of this when it does actually happen and once we know what's going wrong, we can just fix the panic. My guess is that its closing a closed channel or something like that. If we decide that we need some actual signaling to something, it might mean doing something at the TCP level like sending an RST packet or something.

**Comment by max-b** [comment]
> > I'm not super familiar w/ what's going on here. I can see how `recover`ing from the panic here will address the panic, but is there some other way we'd like to be able to be notified that there's been an error? Is that possible? Would it make sense to return an error?
> 
> I was going to do that but then realized that the callers of protocolMainLoop don't actually check the return value. My hope is that we might get some useful logging of this when it does actually happen and once we know what's going wrong, we can just fix the panic. My guess is that its closing a closed channel or something like that. If we decide that we need some actual signaling to something, it might mean doing something at the TCP level like sending an RST packet or something.

Great ok yeah I wasn't clear that the `Printf` would be functionably monitorable. Like I said, I have no about the context, but I wasn't sure that the `stdout` would be logged/monitored. It seems like you have way more sense of all of this, but would it make any more sense to print to `stderr` instead?

**Comment by oxtoacart** [comment]
@max-b Good point. So that we get better visibility into this, I've now shifted to using ops and golog for logging this. That'll allow us to track the error rate and specific errors in borda, and see the log in our usual format.

It's weird integrating our libraries into a fork of netstack, but netstack ceased development a while back so we won't be contributing back upstream anyway.

**Comment by max-b** [comment]
> It's weird integrating our libraries into a fork of netstack, but netstack ceased development a while back so we won't be contributing back upstream anyway.

Yeah I agree that it's a little strange sticking our observability/logging in there (and in general I find logging in libraries a little strange in go tbh), but this does seem like the most practical way forwards :+1:

File: tcpip/transport/tcp/connect.go [hunk 2]
```diff
@@ -1086,6 +1093,16 @@ func (e *endpoint) disableKeepaliveTimer() {
 // goroutine and is responsible for sending segments and handling received
 // segments.
 func (e *endpoint) protocolMainLoop(handshake bool) *tcpip.Error {
+	op := ops.Begin("netstack_tcp_loop")
+	defer op.End()
```
Comment (max-b) [inline tcpip/transport/tcp/connect.go:1097]: Just for my own personal edification: this won't actually do anything unless some reporter is hooked up to the `ops` right? Or is there some other thing that it does out of the box?

Comment (oxtoacart) [reply tcpip/transport/tcp/connect.go:1097]: Correct, it just for the reporter.

**Comment by oxtoacart** [comment]
> > It's weird integrating our libraries into a fork of netstack, but netstack ceased development a while back so we won't be contributing back upstream anyway.
> 
> Yeah I agree that it's a little strange sticking our observability/logging in there (and in general I find logging in libraries a little strange in go tbh), but this does seem like the most practical way forwards 👍

Yeah, ideally I would expose something on the netstack API to support properly instrumenting it and hook into that from our app, but it doesn't seem worth the effort here.

File: tcpip/transport/tcp/connect.go (hunk 2 above)
Comment (max-b) [reply tcpip/transport/tcp/connect.go:1097]: thanks!

**Comment by oxtoacart** [comment]
Thanks for the review @max-b, I'm going to go ahead and merge.

**Pull request merged**